### ISSUE TITLE
Download a https:// ISO File (Certificate Signed by Unknown Authority)

### ISSUE THREAD

**Issue by lmayorga1980** [hashicorp/packer#4969]
I am trying to provision a new machine with virtualbox.

```
 Error downloading: Get https://<my_company_repository>/rhel-server-7.3-x86_64-dvd.iso: x509: certificate signed by unknown authority
```
Is there a way to disable SSL Verification?. I am able to use `curl` and `wget` without any issues.

**Comment by mwhooker** [comment]
I think it'll work if you use `http` protocol.

We don't support insecure certificates. I'd recommend installing your company's certificates on your machine if you need to use https.

I don't think we have any plans to support insecure tls connections. Is there a reason any of the above won't work?

**Comment by ladar** [comment]
Try:

`curl --insecure --output packer_cache/120acbca7b3d55465eb9f8ef53ad7365f2997d42d4f83d7cc285bf5c71e1131f.iso https://<my_company_repository>/rhel-server-7.3-x86_64-dvd.iso`

Or:

`wget --no-check-certificate --output-document=packer_cache/120acbca7b3d55465eb9f8ef53ad7365f2997d42d4f83d7cc285bf5c71e1131f.iso https://<my_company_repository>/rhel-server-7.3-x86_64-dvd.iso`

And then using a file:// URL in the JSON configuration. 

As for packer, an -insecure flag couldn't hurt. It would be nice if we could use invalid HTTPS certs, and/or create build configurations that reference ISO's without providing a hash. I'm finding that the boot ISOs for Gentoo, Arch, openSUSE and a few others, are updated somewhat frequently, and somewhat randomly causing build configurations to break randomly. 

I finally wrote a shell script so I can very the hashes URLs and/or hashes before kicking off a build so the process doesn't fail in the middle of the night. Even then, I've had the ISOs get updated during the window between building the same machine using a different provider. If anyone else [wants it:](https://gist.github.com/ladar/8d188ee3324939e6bd7bae1c8ea9d145)

```bash
#!/bin/bash

# Name: links.sh
# Author: Ladar Levison
#
# Description: Used to verify packer configurations, and validate URLs before kicking off a build.

# The list of packer config files.
FILES="packer-docker.json packer-libvirt.json packer-vmware.json packer-virtualbox.json"

# Collect the list of ISO urls.
ISOURLS=(`grep --no-messages -E "iso_url|guest_additions_url" ${FILES} | awk -F'"' '{print $4}'`)
ISOSUMS=(`grep --no-messages -E "iso_checksum|guest_additions_sha256" ${FILES} | grep -v "iso_checksum_type" | awk -F'"' '{print $4}'`)

# Verify all of the ISO locations are still valid.
function verify_url {

  # Grab just the response header and look for the 200 response code to indicate the link is valid.
  curl --silent --head "$1" | head -1 | grep --silent --extended-regexp "HTTP/1\.1 200 OK|HTTP/2\.0 200 OK"

  # The grep return code tells us whether it found a match in the header or not.
  if [ $? != 0 ]; then
    printf "Link Failure:  $1\n\n"
    exit 1
  fi
}

# Verify all of the ISO locations are valid and then download the ISO and verify the hash.
function verify_sum {

  # Grab just the response header and look for the 200 response code to indicate the link is valid.
  curl --silent --head "$1" | head -1 | grep --silent --extended-regexp "HTTP/1\.1 200 OK|HTTP/2\.0 200 OK"

  # The grep return code tells us whether it found a match in the header or not.
  if [ $? != 0 ]; then
    printf "Link Failure:  $1\n\n"
    exit 1
  fi

  # Grab the ISO and pipe the data through sha256sum, then compare the checksum value.
  curl --silent "$1" | sha256sum | grep --silent "$2"

  # The grep return code tells us whether it found a match in the header or not.
  if [ $? != 0 ]; then
    SUM=`curl --silent "$1" | sha256sum | awk -F' ' '{print $1}'`
    printf "Hash Failure:  $1\n"
    printf "Found       -  $SUM\n"
    printf "Expected    -  $SUM\n\n"
    exit 1
  fi

  printf "Validated   :  $1\n"
  return 0
}

# Validate the templates before building.
function verify_json() {
  packer validate $1
  if [[ $? != 0 ]]; then
    tput setaf 1; tput bold; printf "\n\nthe $1 packer template failed to validate...\n\n"; tput sgr0
    for i in 1 2 3; do printf "\a"; sleep 1; done
    exit 1
  fi
}

function links() {

  for ((i = 0; i < ${#ISOURLS[@]}; ++i)); do
      verify_url "${ISOURLS[$i]}" "${ISOSUMS[$i]}"
  done

  # Let the user know all of the links passed.
  printf "\nAll ${#ISOURLS[@]} of the install media locations are still valid...\n\n"
}

function sums() {

  for ((i = 0; i < ${#ISOURLS[@]}; ++i)); do
      verify_sum "${ISOURLS[$i]}" "${ISOSUMS[$i]}"
  done

  # Let the user know all of the links passed.
  printf "\nAll ${#ISOURLS[@]} of the install media locations are still valid...\n\n"
}

function validate() {

  LIST=($FILES)

  for ((i = 0; i < ${#LIST[@]}; ++i)); do
      verify_json "${LIST[$i]}"
  done

}

# The generic functions.
if [[ $1 == "links" ]]; then links
elif [[ $1 == "sums" ]]; then sums
elif [[ $1 == "validate" ]]; then validate

# Catchall
else
	echo ""
	echo " Commands"
	echo $"   `basename $0` {sums|links|validate}"
	echo ""
	echo " Packer Files"
	echo $"   $FILES"
	echo ""
	echo " Please select a target and run this command again."
	echo ""
	exit 2
fi
```

**Comment by lmayorga1980** [comment]
@ladar ,
thanks for the script and i guess my CI config can pull the ISO in a different process and use my Packer template with the **file://** option but I was just wondering if vagrant has a https://www.vagrantup.com/docs/vagrantfile/machine_settings.html (download_insecure) why doesn't Packer also support this as well?

**Comment by ladar** [comment]
If I had to guess... because vagrant has been around longer, and is frequently used to pull boxes off internal repos. Packer on the other hand presumes that if your downloading the ISO via a network, it needs to verify all that it can, and this the strict hash requirement + cert validation. My guess is this... the developers presumed you'll either a) use HTTP and rely solely on the hash, b) download the ISO separately, and reference it using a file:// link, or c) setup your own HTTPS server with a valid cert (you could use a free Let's Encrypt cert, or set it up with a private CA which you have installed on your machine). I know a number of people employ one of the latter two strategies for building non-public systems like Winodws / MacOS / RHEL, etc - since those ISOs aren't readily available via anonymous public websites.

Personally, I understand your pain. I've spent quite a bit of time trying to ensure all of my images, spanning 16 different distros at this point, download every ISO via HTTPS (vs HTTP),, and know that finding a mirror with a valid HTTPS certificate can often be a bit tricky. It seems many of the smaller projects have mirrors which either don't offer valid HTTPS certificates, or they rely on redirects to CDNs which yield CN mismatches. 

Long story short, I've managed to get HTTPS for everything at this point... except the Virtualbox guest additions ISO. I can't seem to find an official mirror for [this](http://download.virtualbox.org/virtualbox/5.1.22/), as the HTTPS version of that URL will yield a CN mismatch. I also have trouble with the OpenBSD installer timing out when you use HTTPS - but that isn't really a packer issue.

**Comment by lmayorga1980** [comment]
Agreed. I don't think this is an issue. closing it.
